src/lib.rs

```diff
@@ -36,47 +37,60 @@ extern crate bytes;
 extern crate futures;
 extern crate futures_cpupool;
 
-use std::{fmt, fs, io};
+use std::{fmt, fs};
 use std::path::Path;
 
-use futures::{Future, Poll};
-use futures_cpupool::{CpuFuture, CpuPool};
+use futures::{Async, Future, Poll};
+use futures::future::{lazy, Executor};
+use futures::sync::oneshot::{self, Receiver};
+use futures_cpupool::CpuPool;
 
+pub use self::error::Error;
 pub use self::read::{FsReadStream, ReadOptions};
 pub use self::write::{FsWriteSink, WriteOptions};
 
+mod error;
 mod read;
 mod write;
 
 /// A pool of threads to handle file IO.
 #[derive(Clone)]
-pub struct FsPool {
-    cpu_pool: CpuPool,
+pub struct FsPool<E>
+where
+    E: Executor<Box<Future<Item = (), Error = ()> + Send>> + Clone,
```

Comment:
Instead of a generic, I was thinking more-so that `FsPool` could own an internal type that is either `CpuPool`, or an `Executor` trait object. (Or perhaps even just the trait object...) The reasons are these:

- There is a user cost to having a generic. It can become a speed bump to then name either an `FsPool<WhatIsMyExecutor>`, or the streams from it.
- While a trait object means dynamic dispatch, that likely is nothing compared to sending a message to another thread, or doing disk operations.

Reply:
I've altered FsPool to use a trait object instead of a generic.

Reply:
I'm also wondering what the benefit of a union type of an Executor and a CpuPool would be, since a CpuPool is an executor.